libpod/runtime_img.go

```diff
@@ -76,6 +80,79 @@ type ImageFilterParams struct {
 	ImageInput       string
 }
 
+// struct for when a user passes a short or incomplete
+// image name
+type imagePullStruct struct {
+	imageName   string
+	tag         string
+	registry    string
+	hasRegistry bool
+	transport   string
+}
+
+func (ips imagePullStruct) returnFQName() string {
+	return fmt.Sprintf("%s%s/%s:%s", ips.transport, ips.registry, ips.imageName, ips.tag)
+}
+
+func getRegistriesToTry(image string) ([]string, error) {
+	var registries []string
+	var imageError = fmt.Sprintf("unable to parse '%s'\n", image)
+	imgRef, err := reference.Parse(image)
+	if err != nil {
+		return nil, errors.Wrapf(err, imageError)
+	}
+	tagged, isTagged := imgRef.(reference.NamedTagged)
+	tag := "latest"
+	if isTagged {
+		tag = tagged.Tag()
+	}
+	hasDomain := true
+	registry := reference.Domain(imgRef.(reference.Named))
+	if registry == "" {
+		hasDomain = false
+	}
+	imageName := reference.Path(imgRef.(reference.Named))
+	pImage := imagePullStruct{
+		imageName,
+		tag,
+		registry,
+		hasDomain,
+		"docker://",
+	}
+	if pImage.hasRegistry {
+		// If input has a registry, we have to assume they included an image
+		// name but maybe not a tag
+		_, err := alltransports.ParseImageName(pImage.returnFQName())
+		if err != nil {
+			return nil, errors.Errorf(imageError)
+		}
+		registries = append(registries, pImage.returnFQName())
+	} else {
+		// No registry means we check the globals registries configuration file
+		// and assemble a list of candidate sources to try
+		registryConfigPath := ""
+		envOverride := os.Getenv("REGISTRIES_CONFIG_PATH")
+		if len(envOverride) > 0 {
+			registryConfigPath = envOverride
+		}
+		searchRegistries, err := sysregistries.GetRegistries(&types.SystemContext{SystemRegistriesConfPath: registryConfigPath})
+		if err != nil {
+			fmt.Println(err)
+			return nil, errors.Errorf("unable to parse the registries.conf file and"+
+				" the image name '%s' is incomplete.", imageName)
+		}
+		for _, searchRegistry := range searchRegistries {
+			pImage.registry = searchRegistry
+			_, err := alltransports.ParseImageName(pImage.returnFQName())
```

Comment:
While I appreciate trying to insert the registry host name by editing the components with their named semantics, the `reference` package does not make it easy, and doing it correctly (with `:tag`, `@digest`, and _both_ `:tag@digest`) is difficult for little benefit.

As this moment a plain `docker.ParseReference(searchRegistry + `/` + image)` would probably be simpler. (It _would_ be possible to teach `returnFQName()` the [tag][digest] formatting, but my personal aesthetic preference would be to do the string concatenation in this one case because it’s so much simpler.)